### DISCUSSION TITLE
OpenGL Context customization and MSAA support with Avalonia 11.0.4

### DISCUSSION THREAD

**Discussion by jmb10101** [AvaloniaUI/Avalonia#12837] [Q&A] [answered]
I have just upgraded my Avalonia application to 11.0.4 and noticed there were some changes to the OpenGL context that is set up through the OpenGlControlBase initialization. Is there a mechanism to switch between OpenGL versions, or is OpenGL ES required for using this control?
I would also like to enable MSAA for the GL context, but from what I can tell the context that is set up does not support multisampling. Is this true? How do I check? In the previous version, I was able to use OpenGL4 and set up multisampled textures to perform to anti-aliasing, but the bindings I need don't seem to exist in OpenGL ES. What options do I have to enable MSAA?

**Comment by kekekeks** [comment, marked as answer]
OpenGlControlBase picks the context that is +- guaranteed to work. 

You could use [GPU interop](https://github.com/AvaloniaUI/Avalonia/issues/9925) API to import surfaces from your own OpenGL context (in case of Windows via DXGI handle sharing).

**Reply by kekekeks** [reply]
You can also switch to WGL instead of Angle, it should give you full OpenGL on Windows. Use Win32RenderingMode.Wgl in Win32PlatformOptions

**Reply by ArtjomP** [reply]
I have the following options on Avalonia 11.0.6 on Windows:
```cs
.With(new Win32PlatformOptions
{
    RenderingMode = new List<Win32RenderingMode>
    {
        Win32RenderingMode.Wgl
    },
    WglProfiles = new List<GlVersion>
    {
        new GlVersion(GlProfileType.OpenGL, 4, 0)
    }
})
```

and get the next error:
```
Exception Info: System.InvalidOperationException: Win32PlatformOptions.RenderingMode has a value of "Wgl", but no options were applied.
   at Avalonia.Win32.Win32GlManager.InitializeCore()
   at Avalonia.Win32.Win32GlManager.Initialize()
   at Avalonia.Win32.Win32Platform.Initialize(Win32PlatformOptions options)
   at Avalonia.Win32ApplicationExtensions.<>c.<UseWin32>b__0_0()
   at Avalonia.AppBuilder.SetupUnsafe()
   at Avalonia.AppBuilder.Setup()
   at Avalonia.AppBuilder.SetupWithLifetime(IApplicationLifetime lifetime)
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime(AppBuilder builder, String[] args, ShutdownMode shutdownMode)
   at EMU.Executable.Main(String[] args)
```

What options I've missed here? 

Thank you!

**Reply by jp2masa** [reply]
It means that WGL failed to initialize here:

https://github.com/AvaloniaUI/Avalonia/blob/4a30a546fa17c95a06ec0da76c0b67707088a986/src/Windows/Avalonia.Win32/Win32GlManager.cs#L59
